src/validation.rs

```diff
@@ -0,0 +1,320 @@
+use std::error::Error;
+use std::fmt;
+use url::{self, Url};
+
+// -- Macros --
+
+/// Unwrap a Result<T,E>. Early return an Option<E> if unwrapping fails.
+macro_rules! tryopt {
+    ($expr:expr) => (match $expr {
+        Ok(val) => val,
+        Err(err) => return Some(::std::convert::From::from(err)),
+    })
+}
+
+// -- Errors --
+
+/// Union type for errors encountered during validation
+#[derive(Debug)]
+pub enum ValidationError {
+    Parse(url::ParseError),
+    MismatchedOrigin,
+    NotBareOrigin,
+    BadScheme,
+    UserinfoPresent,
+    InconsistentSerialization,
+    BadPort,
+}
+
+impl fmt::Display for ValidationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ValidationError::Parse(ref err) => err.fmt(f),
+            ValidationError::MismatchedOrigin => write!(f, "origins did not match"),
+            ValidationError::NotBareOrigin => write!(f, "uri was not a bare origin"),
+            ValidationError::BadScheme => write!(f, "scheme was not http(s)"),
+            ValidationError::UserinfoPresent => write!(f, "a username or password was specified in the uri"),
+            ValidationError::InconsistentSerialization => write!(f, "parsing and re-serializing the uri changed its representation, check for unnecessary information like default ports"),
+            ValidationError::BadPort => write!(f, "invalid port specified"),
+        }
+    }
+}
+
+impl Error for ValidationError {
+    fn description(&self) -> &str {
+        match *self {
+            ValidationError::Parse(ref err) => err.description(),
+            ValidationError::MismatchedOrigin => "origins did not match",
+            ValidationError::NotBareOrigin => "uri was not a bare origin",
+            ValidationError::BadScheme => "scheme was not http(s)",
+            ValidationError::UserinfoPresent => "a username or password was specified in the uri",
+            ValidationError::InconsistentSerialization => "parsing and re-serializing the uri changed its representation, check for unnecessary information like default ports",
+            ValidationError::BadPort => "invalid port specified",
+        }
+    }
+
+    fn cause(&self) -> Option<&Error> {
+        match *self {
+            ValidationError::Parse(ref err) => Some(err),
+            ValidationError::MismatchedOrigin
+                | ValidationError::NotBareOrigin
+                | ValidationError::BadScheme
+                | ValidationError::UserinfoPresent
+                | ValidationError::InconsistentSerialization
+                | ValidationError::BadPort
+                => None,
+        }
+    }
+}
+
+impl From<url::ParseError> for ValidationError {
+    fn from(err: url::ParseError) -> ValidationError {
+        ValidationError::Parse(err)
+    }
+}
+
+// -- Validation Functions --
+
+/// Test that a URI is valid and conforms to our expetations.
+pub fn valid_uri(raw_uri: &str) -> Option<ValidationError> {
```

Comment:
`Option<ValidationError>` doesn't feel like an idiomatic return type for this, I think `Result<(), ValidationError>` would be clearer even though it's a bit more typing.